pyro/infer/reparam/strategies.py

```diff
@@ -0,0 +1,180 @@
+        """
+        raise NotImplementedError
+
+
+def _minimal_reparam(fn, is_observed):
+    # Unwrap Independent, Masked, Transformed etc.
+    while hasattr(fn, "base_dist"):
+        if isinstance(fn, torch.distributions.TransformedDistribution):
+            if _minimal_reparam(fn.base_dist, is_observed) is None:
+                return None  # No need to reparametrize.
+            else:
+                return TransformReparam()  # Then reparametrize new sites.
+        fn = fn.base_dist
+
+    if isinstance(fn, dist.Stable):
+        if not is_observed:
+            return LatentStableReparam()
+        elif fn.skew.requires_grad or fn.skew.any():
+            return StableReparam()
+        else:
+            return SymmetricStableReparam()
+
+    if isinstance(fn, dist.ProjectedNormal):
+        return ProjectedNormalReparam()
+
+    # TODO apply CircularReparam for VonMises
+
+
+class MinimalReparam(Strategy):
+    """
+    Minimal reparametrization strategy that reparametrizes only those sites
+    that would otherwise lead to error, e.g.
+    :class:`~pyro.distributions.Stable` and
+    :class:`~pyro.distributions.ProjectedNormal` random variables.
+
+    Example::
+
+        @MinimalReparam()
+        def model(...):
+            ...
+
+    which is equivalent to::
+
+        @poutine.reparam(config=MinimalReparam())
+        def model(...):
+            ...
+    """
+
+    def configure(self, msg: dict) -> Optional[Reparam]:
+        return _minimal_reparam(msg["fn"], msg["is_observed"])
+
+
+def _loc_scale_reparam(name, fn):
+    if "_decentered" in name:
+        return  # Avoid infinite recursion.
+
+    # Check for location-scale families.
+    params = set(fn.arg_constraints)
+    if not {"loc", "scale"}.issubset(params):
+        return
+    assert not isinstance(fn, torch.distributions.LogNormal)
+
+    # TODO reparametrize only if parameters are variable. We might guess
+    # based on whether parameters are differentiable, .requires_grad. See
+    # https://github.com/pyro-ppl/pyro/pull/2824
+
+    # Create an elementwise-learnable reparametrizer.
+    shape_params = sorted(params - {"loc", "scale"})
+    return LocScaleReparam(shape_params=shape_params)
+
+
+class AutoReparam(Strategy):
+    """
+    Applies a recommended set of reparametrizers. These currently include:
+    :class:`MinimalReparam`,
+    :class:`~pyro.infer.reparam.transform.TransformReparam`, a fully-learnable
+    :class:`~pyro.infer.reparam.loc_scale.LocScaleReparam`, and
+    :class:`~pyro.infer.reparam.softmax.GumbelSoftmaxReparam`.
+
+    Example::
+
+        @AutoReparam()
+        def model(...):
+            ...
+
+    which is equivalent to::
+
+        @poutine.reparam(config=AutoReparam())
+        def model(...):
+            ...
+
+    .. warning:: This strategy may change behavior across Pyro releases.
+        To inspect or save a given behavior, extract the ``.config`` dict after
+        running the model at least once.
+    """
```

Comment:
TODO add a `centered` argument, so users could use `AutoReparam(centered=0.0)` for MCMC